Add compatibility with Mesa-driven GPUs

10110111 opened the pull request:
This set of changes makes the demo work for on my two different Intel HD Graphics GPUs on Linux driven by Mesa `i965` driver, in addition to nVidia GPU driven by binary `nvidia` driver, which worked before.
To achieve this, the demo is switched to OpenGL Core Profile, framebuffer formats are chosen those which are guaranteed to exist by the OpenGL 3.3 specification, and an unnecessary GLSL extension is unused.

ebruneton commented on Makefile:
```diff
@@ -86,19 +87,22 @@ output/Debug/atmosphere_test: \
 	$(GPP) $^ -o $@
 
 output/Release/atmosphere_integration_test: \
+    output/Debug/glad/src/glad.o \
```
This should be output/Release/

ebruneton commented on Makefile:
```diff
@@ -29,11 +29,11 @@
 GPP_FLAGS := -Wall -Wmain -pedantic -pedantic-errors -std=c++11
 INCLUDE_FLAGS := \
-    -I. -Iexternal -Iexternal/dimensional_types -Iexternal/progress_bar
+    -I. -Iglad/include -Iexternal -Iexternal/dimensional_types -Iexternal/progress_bar
```
Can you move the glad files to external using
  external/glad/include/glad.h
  extenal/glad/src/glad.cc
  extenal/glad/regenerate.sh

ebruneton commented on Makefile:
```diff
@@ -49,6 +49,7 @@ all: lint doc test integration_test demo
 # <regex>.
 lint: $(HEADERS) $(SOURCES)
 	cpplint --exclude=tools/docgen_main.cc \
+            --exclude=glad/src/glad.cc \
```
This can be removed after you move the glad files.

10110111 replied:
This doesn't seem to work. To compile `glad.cc` I have to include `external/glad/src` to `DIRS`. But then `cpplint` tries to check `glad.cc`, which is undesired.

ebruneton commented on atmosphere/demo/demo.cc:
```diff
@@ -136,7 +140,26 @@ Demo::Demo(int viewport_width, int viewport_height) :
+      -1.0, +1.0, 0.0, 1.0,
+      +1.0, +1.0, 0.0, 1.0,
+  };
+  constexpr int coordsPerVertex = 4;
```
Please use kCoordsPerVertex (https://google.github.io/styleguide/cppguide.html#Constant_Names).

ebruneton commented on atmosphere/demo/demo.cc:
```diff
@@ -136,7 +140,26 @@ Demo::Demo(int viewport_width, int viewport_height) :
+  glBindVertexArray(vao_);
+  glGenBuffers(1, &vbo_);
+  glBindBuffer(GL_ARRAY_BUFFER, vbo_);
+  const GLfloat vertices[]= {
```
Missing space between [] and =

ebruneton commented on atmosphere/demo/demo.cc:
```diff
@@ -136,7 +140,26 @@ Demo::Demo(int viewport_width, int viewport_height) :
+  };
+  constexpr int coordsPerVertex = 4;
+  glBufferData(GL_ARRAY_BUFFER, sizeof vertices, vertices, GL_STATIC_DRAW);
+  constexpr GLuint attribIndex = 0;
```
kAttribIndex

ebruneton commented on atmosphere/demo/demo.h:
```diff
@@ -87,6 +90,8 @@ class Demo {
 
   std::unique_ptr<Model> model_;
+  GLuint vao_;
+  GLuint vbo_;
```
Please use more explicit names, for instance fullscreen_quad_vao_ (https://google.github.io/styleguide/cppguide.html#General_Naming_Rules).

ebruneton commented on atmosphere/model.cc:
```diff
@@ -448,11 +447,9 @@ GLuint NewTexture3d(int width, int height, int depth, GLenum format,
+  GLenum internal_format = half_precision ? GL_RGBA16F : GL_RGBA32F;
   glTexImage3D(GL_TEXTURE_3D, 0, internal_format, width, height, depth, 0,
-      format, GL_FLOAT, NULL);
+      GL_RGBA, GL_FLOAT, NULL);
```
I think forcing RGBA is a big drawback here, it increases the GPU memory needs by 33% when this is not always necessary. Could you use 'format' by default and fallback to RGBA only if you get a framebuffer completness error?

ebruneton commented on atmosphere/model.h:
```diff
@@ -322,6 +323,8 @@ class Model {
       bool blend,
       unsigned int num_scattering_orders);
 
+  GLuint vao_;
```
Same comment as above, please use more explicit names.

ebruneton commented on atmosphere/reference/model_test.cc:
```diff
@@ -566,12 +571,32 @@ with the GPU program, and then read back the framebuffer pixels.
-    glVertex4f(+1.0, +1.0, 0.0, 1.0);
-    glEnd();
+    {
+        GLuint vao;
```
Idem.

ebruneton commented on atmosphere/reference/model_test.cc:
```diff
@@ -566,12 +571,32 @@ with the GPU program, and then read back the framebuffer pixels.
+            -1.0, +1.0, 0.0, 1.0,
+            +1.0, +1.0, 0.0, 1.0,
+        };
+        constexpr int coordsPerVertex = 4;
```
kCoordsPerVertex

ebruneton commented on atmosphere/reference/model_test.cc:
```diff
@@ -566,12 +571,32 @@ with the GPU program, and then read back the framebuffer pixels.
+        constexpr int coordsPerVertex = 4;
+        glBufferData(GL_ARRAY_BUFFER, sizeof vertices, vertices,
+                     GL_STATIC_DRAW);
+        constexpr GLuint attribIndex = 0;
```
kAttribIndex

ebruneton commented on text/TextRenderer.cc:
```diff
@@ -0,0 +1,203 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "text/TextRenderer.h"
```
Can you rename TextRenderer.h/.cc to text_renderer.h/.cc for consistency with the other files?

ebruneton commented on text/TextRenderer.cc:
```diff
@@ -0,0 +1,203 @@
+#include "font.inc"
+
+void TextRenderer::setupTexture() {
+    glGenTextures(1, &fontTexture);
```
Please use onyl 2 spaces for indentation (for consistency with the other files).

ebruneton commented on text/TextRenderer.cc:
```diff
@@ -0,0 +1,203 @@
+
+void TextRenderer::setupProgram() {
+    const auto vertexShader = glCreateShader(GL_VERTEX_SHADER);
+    const char*const vertexShaderSrc = R"(#version 330
```
kVertexShaderSrc

ebruneton commented on text/TextRenderer.cc:
```diff
@@ -0,0 +1,203 @@
+    glShaderSource(vertexShader, 1, &vertexShaderSrc, nullptr);
+    glCompileShader(vertexShader);
+    const auto fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+    const char*const fragmentShaderSrc = R"(#version 330
```
kFragmentShaderSrc

ebruneton commented on text/TextRenderer.cc:
```diff
@@ -0,0 +1,203 @@
+
+void TextRenderer::drawChar(char c, int x, int y,
+                            int viewportWidth, int viewportHeight) {
+    if (c<0x20 || c>0x7e) c = '?';
```
Please use spaces around operators, throughout this file ("c < 0x20", "((c >> 4) - 2) / charH", etc)

10110111 replied:
Any way to double-check this? CppLint (on Kubuntu 14.04) seems to be quite happy with current state...

ebruneton commented on text/TextRenderer.h:
```diff
@@ -0,0 +1,55 @@
+#include <glad/glad.h>
+
+class TextRenderer {
+    GLuint fontTexture;
```
Can you make these fields private, use a trailing underscore (https://google.github.io/styleguide/cppguide.html#Variable_Names), and move them after the public members section (https://google.github.io/styleguide/cppguide.html#Declaration_Order)

10110111 replied:
Well they are already private, but OK, I'll move and rename them.

ebruneton commented on text/TextRenderer.h:
```diff
@@ -0,0 +1,55 @@
+    GLuint program;
+    GLfloat color[3];
+
+    void setupTexture();
```
Can you use uppercase for the first letter of these method names (https://google.github.io/styleguide/cppguide.html#Function_Names)?

10110111 commented:
I've addressed your concerns. Would you like me to squash these fixes into the original commits to avoid littering git-log output with style-fix noise etc.?

10110111 commented:
Ah, yeah, I didn't touch the Windows-related files because I can't test them (don't have Windows with its build system handy).

ebruneton commented:
Thanks for your changes, I will squash and merge them. I can't test on Windows either, unfortunately.
